src/sync/markable_arc_cell.rs

```diff
@@ -0,0 +1,244 @@
+use std::marker::PhantomData;
+use std::mem;
+use std::sync::Arc;
+use std::sync::atomic::{AtomicUsize, AtomicBool, Ordering};
+
+/// A `MarkableArcCell` maintains an `Arc<T>` and a marker bit, providing atomic storage and
+/// retrieval of both, as well as atomic `CAS` operations on them.
+#[derive(Debug)]
+pub struct MarkableArcCell<T>(AtomicUsize, AtomicBool, PhantomData<Arc<T>>);
+
+impl<T> Drop for MarkableArcCell<T> {
+    fn drop(&mut self) {
+        self.take();
+    }
+}
+
+impl<T> MarkableArcCell<T> {
+    /// Creates a new `MarkableArcCell` with the specified initial values of the `Arc` and the
+    /// marker bit.
+    pub fn new(t: Arc<T>, m: bool) -> MarkableArcCell<T> {
+        MarkableArcCell(AtomicUsize::new(unsafe { mem::transmute(t) }), AtomicBool::new(m), PhantomData)
+    }
+
+    /// Creates a new `MarkableArcCell` with the specified initial values of the `Arc` interior
+    /// and the marker bit.
+    pub fn with_val(v: T, m: bool) -> MarkableArcCell<T> {
+        MarkableArcCell(AtomicUsize::new(unsafe { mem::transmute(Arc::new(v)) }), AtomicBool::new(m), PhantomData)
+    }
+
+    // Locks the internal spinlock.
+    fn take(&self) -> Arc<T> {
+        loop {
+            match self.0.swap(0, Ordering::Acquire) {
```

Comment:
This isn't lock free, a suspended thread between a take/put operation can block arbitrarily.